File: android/app/src/main/java/com/talentlens/app/engine/ExerciseStateEngine.kt
package com.talentlens.app.engine

import com.talentlens.app.model.BiomechanicsData
import com.talentlens.app.model.ExerciseType
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

enum class FSMWorkoutState {
    IDLE,
    CALIBRATING,
    READY,
    GOING_DOWN,
    BOTTOM_REACHED,
    GOING_UP,
    HOLDING,
    JUMPING,
    PAUSED,
    COMPLETED
}

class ExerciseStateEngine(
    var exerciseType: ExerciseType = ExerciseType.PUSHUPS,
    private val onRepCounted: (Int) -> Unit = {},
    private val onDepthReached: () -> Unit = {},
    private val onFormWarning: (String) -> Unit = {},
    private val onJumpCompleted: (Float) -> Unit = {}
) {
    var state: FSMWorkoutState = FSMWorkoutState.IDLE
        private set

    var score: Int = 0
        private set

    var incompleteCount: Int = 0
        private set

    var currentPrimaryAngle: Float = 180f
        private set

    var currentSecondaryAngle: Float = 180f
        private set

    var depthProgressPercent: Int = 0
        private set

    var formScore: Int = 100
        private set

    var activeSide: String = "right"
        private set

    var peakJumpCm: Float = 0f
        private set

    private var startTimeMs: Long = 0
    private var lastWarningTimeMs: Long = 0
    private val primaryAnglesHistory = mutableListOf<Float>()
    private val secondaryAnglesHistory = mutableListOf<Float>()
    private val repDurationsSec = mutableListOf<Float>()
    private var repStartTimeMs: Long = 0
    private var minAngleInCurrentRep: Float = 180f

    // Jump tracking
    private var jumpTakeoffTimeMs: Long = 0
    private var baselineAnkleY: Float? = null

    // Plank tracking
    private var plankTotalFrames: Int = 0
    private var plankGoodFrames: Int = 0

    fun start() {
        state = FSMWorkoutState.CALIBRATING
        score = 0
        incompleteCount = 0
        formScore = 100
        peakJumpCm = 0f
        startTimeMs = System.currentTimeMillis()
        repStartTimeMs = startTimeMs
        primaryAnglesHistory.clear()
        secondaryAnglesHistory.clear()
        repDurationsSec.clear()
        minAngleInCurrentRep = 180f
        baselineAnkleY = null
        plankTotalFrames = 0
        plankGoodFrames = 0
    }

    fun pause() {
        state = FSMWorkoutState.PAUSED
    }

    fun resume() {
        state = FSMWorkoutState.READY
    }

    fun reset() {
        state = FSMWorkoutState.IDLE
        score = 0
        incompleteCount = 0
        depthProgressPercent = 0
        currentPrimaryAngle = 180f
        currentSecondaryAngle = 180f
    }

    fun finish(elapsedSeconds: Int): BiomechanicsData {
        state = FSMWorkoutState.COMPLETED
        val avgPrimary = if (primaryAnglesHistory.isNotEmpty()) primaryAnglesHistory.average().toFloat() else 85f
        val avgSecondary = if (secondaryAnglesHistory.isNotEmpty()) secondaryAnglesHistory.average().toFloat() else 174f
        val fastestRep = if (repDurationsSec.isNotEmpty()) repDurationsSec.minOrNull() ?: 1.0f else 1.0f
        val finalRpm = if (elapsedSeconds > 0) (score.toFloat() / (elapsedSeconds / 60f)) else 0f
        val finalForm = max(60, 100 - (incompleteCount * 5))

        return BiomechanicsData(
            averageElbowFlexion = if (exerciseType == ExerciseType.PUSHUPS) avgPrimary else 80f,
            averageKneeFlexion = if (exerciseType == ExerciseType.SQUATS) avgPrimary else 85f,
            averageTrunkAlignment = avgSecondary,
            formScore = finalForm,
            incompletedReps = incompleteCount,
            cadenceRpm = finalRpm,
            peakSpeedSec = fastestRep,
            jumpHeightCm = peakJumpCm
        )
    }

    fun processFrame(landmarks: List<Point2D>) {
        if (state == FSMWorkoutState.IDLE || state == FSMWorkoutState.PAUSED || state == FSMWorkoutState.COMPLETED) {
            return
        }
        if (landmarks.size < 33) return

        val leftShoulder = landmarks[11]
        val leftElbow = landmarks[13]
        val leftWrist = landmarks[15]
        val leftHip = landmarks[23]
        val leftKnee = landmarks[25]
        val leftAnkle = landmarks[27]

        val rightShoulder = landmarks[12]
        val rightElbow = landmarks[14]
        val rightWrist = landmarks[16]
        val rightHip = landmarks[24]
        val rightKnee = landmarks[26]
        val rightAnkle = landmarks[28]

        val leftVis = leftShoulder.visibility + leftHip.visibility + leftAnkle.visibility
        val rightVis = rightShoulder.visibility + rightHip.visibility + rightAnkle.visibility

        activeSide = if (rightVis >= leftVis) "right" else "left"

        val shoulder = if (activeSide == "right") rightShoulder else leftShoulder
        val elbow = if (activeSide == "right") rightElbow else leftElbow
        val wrist = if (activeSide == "right") rightWrist else leftWrist
        val hip = if (activeSide == "right") rightHip else leftHip
        val knee = if (activeSide == "right") rightKnee else leftKnee
        val ankle = if (activeSide == "right") rightAnkle else leftAnkle

        val now = System.currentTimeMillis()

        when (exerciseType) {
            // 1. Push-ups
            ExerciseType.PUSHUPS -> {
                val elbowAngle = GeometryUtils.calculateJointAngle(shoulder, elbow, wrist)
                val plankAngle = GeometryUtils.calculatePlankAlignment(shoulder, hip, ankle)

                currentPrimaryAngle = GeometryUtils.smoothAngle(elbowAngle, currentPrimaryAngle)
                currentSecondaryAngle = plankAngle

                primaryAnglesHistory.add(currentPrimaryAngle)
                secondaryAnglesHistory.add(plankAngle)

                val progress = max(0, min(100, (((155f - currentPrimaryAngle) / (155f - 90f)) * 100f).roundToInt()))
                depthProgressPercent = progress

                if (plankAngle < 145f || plankAngle > 200f) {
                    if (now - lastWarningTimeMs > 2500) {
                        onFormWarning("Keep your back straight")
                        lastWarningTimeMs = now
                    }
                }

                when (state) {
                    FSMWorkoutState.CALIBRATING -> {
                        if (currentPrimaryAngle >= 150f) {
                            state = FSMWorkoutState.READY
                        }
                    }
                    FSMWorkoutState.READY -> {
                        if (currentPrimaryAngle < 140f) {
                            state = FSMWorkoutState.GOING_DOWN
                            repStartTimeMs = now
                        }
                    }
                    FSMWorkoutState.GOING_DOWN -> {
                        if (currentPrimaryAngle <= 90f) {
                            state = FSMWorkoutState.BOTTOM_REACHED
                            onDepthReached()
                        } else if (currentPrimaryAngle > 150f) {
                            state = FSMWorkoutState.READY
                            incompleteCount++
                        }
                    }
                    FSMWorkoutState.BOTTOM_REACHED -> {
                        if (currentPrimaryAngle > 105f) {
                            state = FSMWorkoutState.GOING_UP
                        }
                    }
                    FSMWorkoutState.GOING_UP -> {
                        if (currentPrimaryAngle >= 155f) {
                            score++
                            val durSec = (now - repStartTimeMs) / 1000f
                            repDurationsSec.add(durSec)
                            onRepCounted(score)
                            state = FSMWorkoutState.READY
                        }
                    }
                    else -> {}
                }
            }

            // 2. Squats
            ExerciseType.SQUATS -> {
                val kneeAngle = GeometryUtils.calculateKneeFlexion(hip, knee, ankle)
                val hipAngle = GeometryUtils.calculateHipFlexion(shoulder, hip, knee)

                currentPrimaryAngle = GeometryUtils.smoothAngle(kneeAngle, currentPrimaryAngle)
                currentSecondaryAngle = hipAngle

                primaryAnglesHistory.add(currentPrimaryAngle)
                secondaryAnglesHistory.add(hipAngle)

                val progress = max(0, min(100, (((160f - currentPrimaryAngle) / (160f - 90f)) * 100f).roundToInt()))
                depthProgressPercent = progress

                when (state) {
                    FSMWorkoutState.CALIBRATING -> {
                        if (currentPrimaryAngle >= 155f) {
                            state = FSMWorkoutState.READY
                        }
                    }
                    FSMWorkoutState.READY -> {
                        if (currentPrimaryAngle < 145f) {
                            state = FSMWorkoutState.GOING_DOWN
                            repStartTimeMs = now
                        }
                    }
                    FSMWorkoutState.GOING_DOWN -> {
                        if (currentPrimaryAngle <= 90f) {
                            state = FSMWorkoutState.BOTTOM_REACHED
                            onDepthReached()
                        } else if (currentPrimaryAngle > 155f) {
                            state = FSMWorkoutState.READY
                            incompleteCount++
                        }
                    }
                    FSMWorkoutState.BOTTOM_REACHED -> {
                        if (currentPrimaryAngle > 105f) {
                            state = FSMWorkoutState.GOING_UP
                        }
                    }
                    FSMWorkoutState.GOING_UP -> {
                        if (currentPrimaryAngle >= 160f) {
                            score++
                            val durSec = (now - repStartTimeMs) / 1000f
                            repDurationsSec.add(durSec)
                            onRepCounted(score)
                            state = FSMWorkoutState.READY
                        }
                    }
                    else -> {}
                }
            }

            // 3. Plank Hold
            ExerciseType.PLANK -> {
                val plankAngle = GeometryUtils.calculatePlankAlignment(shoulder, hip, ankle)
                currentPrimaryAngle = plankAngle
                primaryAnglesHistory.add(plankAngle)

                plankTotalFrames++
                val isGood = plankAngle in 155f..185f
                if (isGood) plankGoodFrames++

                formScore = if (plankTotalFrames > 0) ((plankGoodFrames.toFloat() / plankTotalFrames) * 100).roundToInt() else 100

                when (state) {
                    FSMWorkoutState.CALIBRATING -> {
                        if (isGood) state = FSMWorkoutState.HOLDING
                    }
                    FSMWorkoutState.HOLDING -> {
                        if (!isGood && now - lastWarningTimeMs > 2500) {
                            onFormWarning(if (plankAngle < 155f) "Hips sagging! Lift core" else "Hips too high")
                            lastWarningTimeMs = now
                        }
                    }
                    else -> {}
                }
            }

            // 4. Vertical Jump
            ExerciseType.VERTICAL_JUMP -> {
                val kneeAngle = GeometryUtils.calculateKneeFlexion(hip, knee, ankle)
                currentPrimaryAngle = kneeAngle

                val ankleY = (leftAnkle.y + rightAnkle.y) / 2f

                when (state) {
                    FSMWorkoutState.CALIBRATING -> {
                        baselineAnkleY = ankleY
                        state = FSMWorkoutState.READY
                    }
                    FSMWorkoutState.READY -> {
                        if (kneeAngle < 125f) {
                            state = FSMWorkoutState.GOING_DOWN
                        }
                    }
                    FSMWorkoutState.GOING_DOWN -> {
                        val base = baselineAnkleY ?: ankleY
                        if (ankleY < base - 0.04f) {
                            state = FSMWorkoutState.JUMPING
                            jumpTakeoffTimeMs = now
                        }
                    }
                    FSMWorkoutState.JUMPING -> {
                        val base = baselineAnkleY ?: ankleY
                        if (ankleY >= base - 0.02f && jumpTakeoffTimeMs > 0) {
                            val flightTime = (now - jumpTakeoffTimeMs) / 1000f
                            if (flightTime in 0.2f..1.0f) {
                                val height = GeometryUtils.calculateJumpHeightFromFlightTime(flightTime)
                                if (height > peakJumpCm) {
                                    peakJumpCm = height
                                    score = height.roundToInt()
                                }
                                onJumpCompleted(height)
                            }
                            state = FSMWorkoutState.READY
                            jumpTakeoffTimeMs = 0
                        }
                    }
                    else -> {}
                }
            }
        }
    }
}

File: ios/TalentLens/Engine/ExerciseStateEngine.swift
//
//  ExerciseStateEngine.swift
//  TalentLens
//
//  Created for TalentLens Olympic-Grade Athletic Testing Suite.
//

import Foundation
import Combine

public enum FSMWorkoutState: String, Codable {
    case idle = "IDLE"
    case calibrating = "CALIBRATING"
    case ready = "READY"
    case goingDown = "GOING_DOWN"
    case bottomReached = "BOTTOM_REACHED"
    case goingUp = "GOING_UP"
    case holding = "HOLDING"
    case jumping = "JUMPING"
    case paused = "PAUSED"
    case completed = "COMPLETED"
}

public class ExerciseStateEngine: ObservableObject {
    public var exerciseType: ExerciseType
    public var onRepCounted: ((Int) -> Void)?
    public var onDepthReached: (() -> Void)?
    public var onFormWarning: ((String) -> Void)?
    public var onJumpCompleted: ((Float) -> Void)?
    
    @Published public private(set) var state: FSMWorkoutState = .idle
    @Published public private(set) var score: Int = 0
    @Published public private(set) var incompleteCount: Int = 0
    @Published public private(set) var currentPrimaryAngle: Float = 180.0
    @Published public private(set) var currentSecondaryAngle: Float = 180.0
    @Published public private(set) var depthProgressPercent: Int = 0
    @Published public private(set) var formScore: Int = 100
    @Published public private(set) var activeSide: String = "right"
    @Published public private(set) var peakJumpCm: Float = 0
    
    private var startTimeMs: Double = 0
    private var lastWarningTimeMs: Double = 0
    private var primaryAnglesHistory: [Float] = []
    private var secondaryAnglesHistory: [Float] = []
    private var repDurationsSec: [Float] = []
    private var repStartTimeMs: Double = 0
    private var minAngleInCurrentRep: Float = 180.0
    
    // Jump tracking
    private var jumpTakeoffTimeMs: Double = 0
    private var baselineAnkleY: Float? = nil
    
    // Plank tracking
    private var plankTotalFrames: Int = 0
    private var plankGoodFrames: Int = 0
    
    public init(
        exerciseType: ExerciseType = .pushups,
        onRepCounted: ((Int) -> Void)? = nil,
        onDepthReached: (() -> Void)? = nil,
        onFormWarning: ((String) -> Void)? = nil,
        onJumpCompleted: ((Float) -> Void)? = nil
    ) {
        self.exerciseType = exerciseType
        self.onRepCounted = onRepCounted
        self.onDepthReached = onDepthReached
        self.onFormWarning = onFormWarning
        self.onJumpCompleted = onJumpCompleted
    }
    
    public func start() {
        state = .calibrating
        score = 0
        incompleteCount = 0
        formScore = 100
        peakJumpCm = 0
        startTimeMs = Date().timeIntervalSince1970 * 1000
        repStartTimeMs = startTimeMs
        primaryAnglesHistory.removeAll()
        secondaryAnglesHistory.removeAll()
        repDurationsSec.removeAll()
        minAngleInCurrentRep = 180.0
        baselineAnkleY = nil
        plankTotalFrames = 0
        plankGoodFrames = 0
    }
    
    public func pause() {
        state = .paused
    }
    
    public func resume() {
        state = .ready
    }
    
    public func reset() {
        state = .idle
        score = 0
        incompleteCount = 0
        depthProgressPercent = 0
        currentPrimaryAngle = 180.0
        currentSecondaryAngle = 180.0
    }
    
    public func finish(elapsedSeconds: Int) -> BiomechanicsData {
        state = .completed
        let avgPrimary = primaryAnglesHistory.isEmpty ? 85.0 : (primaryAnglesHistory.reduce(0, +) / Float(primaryAnglesHistory.count))
        let avgSecondary = secondaryAnglesHistory.isEmpty ? 174.0 : (secondaryAnglesHistory.reduce(0, +) / Float(secondaryAnglesHistory.count))
        let fastestRep = repDurationsSec.isEmpty ? 1.0 : (repDurationsSec.min() ?? 1.0)
        let finalRpm = elapsedSeconds > 0 ? (Float(score) / (Float(elapsedSeconds) / 60.0)) : 0.0
        let finalForm = max(60, 100 - (incompleteCount * 5))
        
        return BiomechanicsData(
            averageElbowFlexion: exerciseType == .pushups ? avgPrimary : 80.0,
            averageKneeFlexion: exerciseType == .squats ? avgPrimary : 85.0,
            averageTrunkAlignment: avgSecondary,
            formScore: finalForm,
            incompletedReps: incompleteCount,
            cadenceRpm: finalRpm,
            peakSpeedSec: fastestRep,
            jumpHeightCm: peakJumpCm
        )
    }
    
    public func processFrame(landmarks: [Point2D]) {
        if state == .idle || state == .paused || state == .completed {
            return
        }
        guard landmarks.count >= 33 else { return }
        
        let leftShoulder = landmarks[11]
        let leftElbow = landmarks[13]
        let leftWrist = landmarks[15]
        let leftHip = landmarks[23]
        let leftKnee = landmarks[25]
        let leftAnkle = landmarks[27]
        
        let rightShoulder = landmarks[12]
        let rightElbow = landmarks[14]
        let rightWrist = landmarks[16]
        let rightHip = landmarks[24]
        let rightKnee = landmarks[26]
        let rightAnkle = landmarks[28]
        
        let leftVis = leftShoulder.visibility + leftHip.visibility + leftAnkle.visibility
        let rightVis = rightShoulder.visibility + rightHip.visibility + rightAnkle.visibility
        
        activeSide = (rightVis >= leftVis) ? "right" : "left"
        
        let shoulder = (activeSide == "right") ? rightShoulder : leftShoulder
        let elbow = (activeSide == "right") ? rightElbow : leftElbow
        let wrist = (activeSide == "right") ? rightWrist : leftWrist
        let hip = (activeSide == "right") ? rightHip : leftHip
        let knee = (activeSide == "right") ? rightKnee : leftKnee
        let ankle = (activeSide == "right") ? rightAnkle : leftAnkle
        
        let now = Date().timeIntervalSince1970 * 1000
        
        switch exerciseType {
        case .pushups:
            let elbowAngle = GeometryUtils.calculateJointAngle(a: shoulder, b: elbow, c: wrist)
            let plankAngle = GeometryUtils.calculatePlankAlignment(shoulder: shoulder, hip: hip, ankle: ankle)
            
            currentPrimaryAngle = GeometryUtils.smoothAngle(current: elbowAngle, previous: currentPrimaryAngle)
            currentSecondaryAngle = plankAngle
            
            primaryAnglesHistory.append(currentPrimaryAngle)
            secondaryAnglesHistory.append(plankAngle)
            
            let progress = max(0, min(100, Int(((155.0 - currentPrimaryAngle) / (155.0 - 90.0)) * 100.0)))
            depthProgressPercent = progress
            
            if plankAngle < 145.0 || plankAngle > 200.0 {
                if now - lastWarningTimeMs > 2500 {
                    onFormWarning?("Keep your back straight")
                    lastWarningTimeMs = now
                }
            }
            
            switch state {
            case .calibrating:
                if currentPrimaryAngle >= 150.0 {
                    state = .ready
                }
            case .ready:
                if currentPrimaryAngle < 140.0 {
                    state = .goingDown
                    repStartTimeMs = now
                }
            case .goingDown:
                if currentPrimaryAngle <= 90.0 {
                    state = .bottomReached
                    onDepthReached?()
                } else if currentPrimaryAngle > 150.0 {
                    state = .ready
                    incompleteCount += 1
                }
            case .bottomReached:
                if currentPrimaryAngle > 105.0 {
                    state = .goingUp
                }
            case .goingUp:
                if currentPrimaryAngle >= 155.0 {
                    score += 1
                    let durSec = Float((now - repStartTimeMs) / 1000.0)
                    repDurationsSec.append(durSec)
                    onRepCounted?(score)
                    state = .ready
                }
            default:
                break
            }
            
        case .squats:
            let kneeAngle = GeometryUtils.calculateKneeFlexion(hip: hip, knee: knee, ankle: ankle)
            let hipAngle = GeometryUtils.calculateHipFlexion(shoulder: shoulder, hip: hip, knee: knee)
            
            currentPrimaryAngle = GeometryUtils.smoothAngle(current: kneeAngle, previous: currentPrimaryAngle)
            currentSecondaryAngle = hipAngle
            
            primaryAnglesHistory.append(currentPrimaryAngle)
            secondaryAnglesHistory.append(hipAngle)
            
            let progress = max(0, min(100, Int(((160.0 - currentPrimaryAngle) / (160.0 - 90.0)) * 100.0)))
            depthProgressPercent = progress
            
            switch state {
            case .calibrating:
                if currentPrimaryAngle >= 155.0 {
                    state = .ready
                }
            case .ready:
                if currentPrimaryAngle < 145.0 {
                    state = .goingDown
                    repStartTimeMs = now
                }
            case .goingDown:
                if currentPrimaryAngle <= 90.0 {
                    state = .bottomReached
                    onDepthReached?()
                } else if currentPrimaryAngle > 155.0 {
                    state = .ready
                    incompleteCount += 1
                }
            case .bottomReached:
                if currentPrimaryAngle > 105.0 {
                    state = .goingUp
                }
            case .goingUp:
                if currentPrimaryAngle >= 160.0 {
                    score += 1
                    let durSec = Float((now - repStartTimeMs) / 1000.0)
                    repDurationsSec.append(durSec)
                    onRepCounted?(score)
                    state = .ready
                }
            default:
                break
            }
            
        case .plank:
            let plankAngle = GeometryUtils.calculatePlankAlignment(shoulder: shoulder, hip: hip, ankle: ankle)
            currentPrimaryAngle = plankAngle
            primaryAnglesHistory.append(plankAngle)
            
            plankTotalFrames += 1
            let isGood = (plankAngle >= 155.0 && plankAngle <= 185.0)
            if isGood { plankGoodFrames += 1 }
            
            formScore = plankTotalFrames > 0 ? Int((Float(plankGoodFrames) / Float(plankTotalFrames)) * 100.0) : 100
            
            switch state {
            case .calibrating:
                if isGood { state = .holding }
            case .holding:
                if !isGood && (now - lastWarningTimeMs > 2500) {
                    onFormWarning?(plankAngle < 155.0 ? "Hips sagging! Lift core" : "Hips too high")
                    lastWarningTimeMs = now
                }
            default:
                break
            }
            
        case .verticalJump:
            let kneeAngle = GeometryUtils.calculateKneeFlexion(hip: hip, knee: knee, ankle: ankle)
            currentPrimaryAngle = kneeAngle
            
            let ankleY = Float((leftAnkle.y + rightAnkle.y) / 2.0)
            
            switch state {
            case .calibrating:
                baselineAnkleY = ankleY
                state = .ready
            case .ready:
                if kneeAngle < 125.0 {
                    state = .goingDown
                }
            case .goingDown:
                let base = baselineAnkleY ?? ankleY
                if ankleY < base - 0.04 {
                    state = .jumping
                    jumpTakeoffTimeMs = now
                }
            case .jumping:
                let base = baselineAnkleY ?? ankleY
                if ankleY >= base - 0.02 && jumpTakeoffTimeMs > 0 {
                    let flightTime = Float((now - jumpTakeoffTimeMs) / 1000.0)
                    if flightTime >= 0.2 && flightTime <= 1.0 {
                        let height = GeometryUtils.calculateJumpHeightFromFlightTime(flightTimeSec: flightTime)
                        if height > peakJumpCm {
                            peakJumpCm = height
                            score = Int(height.rounded())
                        }
                        onJumpCompleted?(height)
                    }
                    state = .ready
                    jumpTakeoffTimeMs = 0
                }
            default:
                break
            }
        }
    }
}
